programs/stake/src/stake_state.rs

```diff
@@ -856,6 +856,113 @@ pub fn merge(
+        ic_msg!(
+            invoke_context,
+            "expected owner {}, got {}",
+            id(),
+            *uninitialized_stake_account.get_owner()
+        );
+        return Err(InstructionError::IncorrectProgramId);
+    }
+    if uninitialized_stake_account.get_data().len() != StakeState::size_of() {
+        ic_msg!(
+            invoke_context,
+            "expected len {}, got len {}",
+            StakeState::size_of(),
+            uninitialized_stake_account.get_data().len()
+        );
+        return Err(InstructionError::InvalidAccountData);
+    }
+    if !matches!(
+        uninitialized_stake_account.get_state()?,
+        StakeState::Uninitialized
+    ) {
+        return Err(InstructionError::InvalidAccountData);
+    }
+
+    // validate the provided vote account
+    let vote_account = instruction_context
+        .try_borrow_instruction_account(transaction_context, vote_account_index)?;
+    if *vote_account.get_owner() != solana_vote_program::id() {
+        return Err(InstructionError::IncorrectProgramId);
+    }
+    let vote_pubkey = *vote_account.get_key();
+    let vote_state = vote_account.get_state::<VoteStateVersions>();
+
+    // ensure `stake_account` is active, check for presence of the stake authority, and
+    // extract its Meta
+    let meta = if let StakeState::Stake(meta, stake) = stake_account.get_state()? {
+        meta.authorized.check(signers, StakeAuthorize::Staker)?;
+
+        let status = stake
+            .delegation
+            .stake_activating_and_deactivating(clock.epoch, Some(stake_history));
+        if status.effective == 0 || status.activating != 0 || status.deactivating != 0 {
+            ic_msg!(invoke_context, "stake is not active");
+            return Err(InstructionError::InvalidAccountData);
+        }
+
+        // Deny redelegating to the same vote account. This is nonsensical and could be used to
+        // grief the global stake warm-up/cool-down rate
+        if stake.delegation.voter_pubkey == vote_pubkey {
+            ic_msg!(
+                invoke_context,
+                "redelegating to the same vote account not permitted"
+            );
+            return Err(InstructionError::InvalidArgument);
+        }
+
+        meta
+    } else {
+        return Err(InstructionError::InvalidAccountData);
+    };
+
+    // deactivate `stake_account`
+    deactivate(stake_account, clock, signers)?;
+
+    // transfer all but the `rent_exempt_reserve` balance to the uninitialized stake account
+    let redelegation_amount = stake_account
+        .get_lamports()
+        .checked_sub(meta.rent_exempt_reserve)
+        .ok_or(InstructionError::InsufficientFunds)?;
+    stake_account.checked_sub_lamports(redelegation_amount)?;
+    uninitialized_stake_account.checked_add_lamports(redelegation_amount)?;
+
+    // initialize and delegate `uninitialized_stake_account`
+    let ValidatedDelegatedInfo { stake_amount } = validate_delegated_amount(
```

Comment:
This could be confusing, since if you have an uninitialized stake account with 1000 SOL, and you redelegate from an account with 1 SOL, you'd activate that whole amount.  This is technically fine, since it's similar to activating one stake account with 1000 SOL and also redelegating that 1 SOL to the same validator.  For clarity though, I'd prefer that only the existing stake account's delegation is redelegated.  What do you think?

Reply:
I actually very intentionally left this option it after thinking about it for a while. There's no risk that I see and it enables one to redelegate **and** add additional stake in one pass, which could be enable some stake pool implementations to reduce the number of transient stake accounts they require.